platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,438 @@
+documents for Gossip.
+
+![Gossip](gossip-system.png)
+
+Gossip is the only part of Consensus that communicates over the network with gossip peers. When Gossip is initialized,
+it is supplied a "roster". This roster contains the full set of nodes participating in gossip, along with their metadata
+such as RSA signing keys, IP addresses, and so forth. The Gossip module decides which peers to gossip with (using
+whatever algorithm it chooses).
+
+#### Events
+
+Gossip is event-oriented, meaning that it is given events to gossip, and emits events it receives through gossip. An
+implementation of Gossip could be based on a lower-level implementation based on bytes, but at the module level, it
+works in terms of events.
+
+When the Gossip module receives events through gossip, it *may* choose to perform some deduplication before sending
+them to Event Intake, but it is not required to do so.
+
+Honest nodes will **only** send valid events in *topological order* to peers. A peer may still receive events out of
+order, because different events may arrive from different peers at different times. To support topological ordering,
+events received by Gossip are not retained by Gossip, they are simply passed directly to Event Intake, which after
+verification, also orders events in topological order and recorded durably to disk, and then send back to the Gossip
+module to be sent out to its gossip peers.
+
+**The key design principle here is that a node will only send valid events through gossip**. If invalid events are
+ever received, you may know the node that sent them to you is dishonest. Validating events increases latency at each
+"hop", but allows us to identify dishonest gossip peers and discipline them accordingly.
+
+During execution, for all nodes that are online and able to keep up, events are received "live" and processed
+immediately and re-gossiped. However, if a node is offline and then comes back online, or is starting back up after
+reconnect, it may be missing events. In this case, the node will need to ask its peers for any events it is missing.
+
+For this reason, every honest node needs to buffer some events, so when its peer asks it for events, it is able to
+give them. Events can be "non-ancient", "ancient", or "expired". Although the definition for these is subject to
+configuration, we will here use some example values. Non-ancient events are those events that are either not yet
+in a consensus round, or are in one of the last 26 consensus rounds. Expired events are those that are in a round
+that is older than the last most recent 500 consensus rounds. And ancient events are those that are in a round between
+26 and 500.
+
+The Gossip module must cache all non-expired events. Note that it must be a fixed number of rounds that defines
+"expired", because otherwise it is possible for a quiescent network to never recover (see the documentation for
```

Comment:
I don't think expired events are a necessary ingredient for the network to recover, although they are definitely needed for a robust and highly performant network. It's possible for the tipset algorithm to always make progress with only non-ancient events and reconnect (for the nodes that are missing events that are ancient to everybody else).

Reply:
Except reconnect won't work if the number is too small. "500" is a proxy for "large enough that when you reconnect you can then catch up whatever you missed through gossip". If nodes didn't keep that 500, then you couldn't catch up.

We also need *something* beyond just caching non-ancient events, otherwise if I miss just a single event that, by the time I ask for it, my peers don't have it, then I have to go into reconnect (no doubt what you mean by "robust and highly performant").

Reply:
I agree, quibbling over semantics. 😜